numpy/core/src/multiarray/nditer_pywrap.c

```diff
@@ -2486,9 +2488,33 @@ npyiter_enter(NewNpyArrayIterObject *self)
         PyErr_SetString(PyExc_ValueError, "operation on  non-initialized iterator");
         return NULL;
     }
-    if (self->managed != CONTEXT_NOTENTERED) {
-        PyErr_SetString(PyExc_ValueError, "cannot reenter managed nditer context");
-        return NULL;
+    if (self->managed == CONTEXT_EXITED && self->needs_context_manager) {
+        /* Re-entering context manager after already exiting */
+        PyArrayObject ** object;
+        int iop, nop;
+        nop = NpyIter_GetNOp(self->iter);
+        object = NpyIter_GetOperandArray(self->iter);
+        iop = 0;
+        for (iop = 0; iop < nop; ++iop) {
+            if (self->writebackifcopy[iop] != 0) {
+                /* Reconstruct the WRITEBACKIFCOPY flag */
+                int ret;
+                PyObject *base = ((PyArrayObject_fields *)object[iop])->base;
+                PyObject *temp = PyArray_NewLikeArray(object[iop],
```

Comment:
Is there another way to (re)allocate the data buffer on the ``object[iop]``? It was allocated in the nditer constructor, deleted in ``PyArray_ResolveWritebackIfCopy`` during ``npyiter_exit``. Now we have re-entered the context manager and it must be recreated. Replacing the original ``object[iop]`` with temp on line 2515 is ugly.

Reply:
Just from a quick look, `npy_alloc_cache` may be useful.